### PR TITLE
Fix merging with conflict by upgrade mergo version to v0.3.12

### LINKED ISSUE #107499: remove use of github.com/imdario/mergo (currently use is incompatible with latest mergo version, latest mergo version introduces bugs)
### What happened?

client-go behavior changes depending on version of github.com/imdario/mergo used. Kubernetes pins to `v0.3.5`. Many other projects in the ecosystem, including Kubernetes projects like sigs.k8s.io/controller-runtime, use newer versions like `0.3.12`.

If a consumer of client-go is not careful, they will very likely automatically use `mergo`. This will subtly give different behavior than client-go declares. Users of client-go must therefor use a `replace` for mergo (if they care about this issue), which also implies all downstream consumers must do the same and `go install` will not work.

### What did you expect to happen?

client-go behavior is the same regardless of mergo version.

### How can we reproduce it (as minimally and precisely as possible)?

```go
package kube

import (
	"fmt"
	"os"
	"path/filepath"
	"testing"

	"k8s.io/client-go/tools/clientcmd"
)

func TestMergo(t *testing.T) {
	config1, err := mkKubeConfig("1.1.1.1", "3.3.3.3")
	if err != nil {
		t.Fatalf("Failed to create a sample kubernetes config file. Err: %v", err)
	}
	defer os.RemoveAll(filepath.Dir(config1))
	config2, err := mkKubeConfig("2.2.2.2", "4.4.4.4")
	if err != nil {
		t.Fatalf("Failed to create a sample kubernetes config file. Err: %v", err)
	}
	defer os.RemoveAll(filepath.Dir(config2))

	if err := os.Setenv("KUBECONFIG", config1+":"+config2); err != nil {
		t.Fatalf("Failed to set KUBECONFIG environment variable")
	}

	loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
	loadingRules.DefaultClientConfig = &clientcmd.DefaultClientConfig
	configOverrides := &clientcmd.ConfigOverrides{
		ClusterDefaults: clientcmd.ClusterDefaults,
	}

	//resp, err := BuildClientConfig("", "")
	resp, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, configOverrides).ClientConfig()
	if err != nil {
		t.Fatal(err)
	}
	wantHost := "https://1.1.1.1:8001"
	if resp != nil && resp.Host != wantHost {
		t.Fatalf("Incorrect host. Got: %s, Want: %s", resp.Host, wantHost)
	}
}

func mkKubeConfig(cluster1Host string, cluster2Host string) (string, error) {
	tempDir, err := os.MkdirTemp("/tmp/", ".kube")
	if err != nil {
		return "", err
	}
	filePath := filepath.Join(tempDir, "config")

	template := `apiVersion: v1
kind: Config
clusters:
- cluster:
    insecure-skip-tls-verify: true
    server: https://%s:8001
  name: cluster.local
- cluster:
    insecure-skip-tls-verify: true
    server: https://%s:8001
  name: cluster2.local
contexts:
- context:
    cluster: cluster.local
    namespace: default
    user: admin
  name: cluster.local-context
- context:
    cluster: cluster2.local
    namespace: default
    user: admin
  name: cluster2.local-context
current-context: cluster.local-context
preferences: {}
users:
- name: admin
  user:
    token: sdsddsd`

	sampleConfig := fmt.Sprintf(template, cluster1Host, cluster2Host)
	err = os.WriteFile(filePath, []byte(sampleConfig), 0o644)
	if err != nil {
		return "", err
	}
	return filePath, nil
}
```

Run with v0.3.5:
```
=== RUN   TestMergo
--- PASS: TestMergo (0.00s)
PASS
```

Run with v0.3.12:
```
=== RUN   TestMergo
    repro_test.go:41: Incorrect host. Got: https://2.2.2.2:8001, Want: https://1.1.1.1:8001
--- FAIL: TestMergo (0.00s)

FAIL
```

If its easier you can also remove the `replace` in http://github.com/istio/istio and run `go test ./pkg/kube -run TestBuildClientConfig` which will give the same result. The above test is just a simplified version of this test.

### Anything else we need to know?

_No response_

### Kubernetes version

<details>
v0.23.1, but this has been an issue for almost 2 years
</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by jlsong01**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:

- client-go pins to mergo v0.3.5, seems v0.3.5 has some bug, which can not handle merge with confilct correctly.

- Since unit test `Example_mergingSomeWithConflict` uses mergo v0.3.5, the merge process is wrong and the corresponing ut expection is wrong as well.

- We need upgrade mergo version into the latest v0.3.12 and modify the  `Example_mergingSomeWithConflict` expection accordingly.

#### Which issue(s) this PR fixes:

Fixes #107499

#### Special notes for your reviewer:
According to my test, the mergo bug has been fixed since v0.3.8, but have not found the releated issue report.

#### Does this PR introduce a user-facing change?

```release-note
NO
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
NO
```

### CODE REVIEW

**Comment by jlsong01** [comment]
work in progress as some tests failed...

**Comment by jlsong01** [comment]
@kubernetes/dep-approvers Hey, could you please take a look firstly?

**Comment by cblecker** [comment]
/lgtm
/assign @liggitt 

Changes look fine to me, @liggitt.

File: staging/src/k8s.io/client-go/tools/clientcmd/loader_test.go [hunk 1]
```diff
@@ -734,7 +734,8 @@ func Example_mergingSomeWithConflict() {
 	// apiVersion: v1
 	// clusters:
 	// - cluster:
-	//     server: http://cow.org:8080
+	//     insecure-skip-tls-verify: true
+	//     server: http://a-different-cow.org:8080
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/tools/clientcmd/loader_test.go:738]: This unit test is explicitly checking what the resulting configuration is when we merge two files. This test change demonstrates the library update is flipping merge behavior and breaking compatibility with existing kubeconfig merge order.

Revert the changes to this file. If the mergo library changed behavior, we'll need to adjust how we're calling it to preserve the existing behavior of kubeconfig merging.

**Comment by liggitt** [comment]
> Since unit test Example_mergingSomeWithConflict uses mergo v0.3.5, the merge process is wrong and the corresponing ut expection is wrong as well.

That unit test has existed for 6 years. The current behavior of merging two kubeconfig files is what we need to preserve. If we want to update the mergo library, that's ok, but we have to maintain compatibility with *our* existing merge behavior.

**Comment by liggitt** [comment]
/hold
for kubeconfig merging compatibility break

**Comment by liggitt** [comment]
/close

due to inactivity. an update that preserves existing kubeconfig / client-go behavior would be fine

File: vendor/github.com/imdario/mergo/map.go [hunk 2]
```diff
@@ -72,6 +72,7 @@ func deepMap(dst, src reflect.Value, visited map[uintptr]*visit, depth int, conf
 	case reflect.Struct:
 		srcMap := src.Interface().(map[string]interface{})
 		for key := range srcMap {
+			config.overwriteWithEmptyValue = true
```
[the remark this answers is not in the document]
Comment (liggitt) [reply vendor/github.com/imdario/mergo/map.go:75]: yup, reported at https://github.com/imdario/mergo/issues/187

[the remark this answers is not in the document]
Comment (jlsong01) [reply vendor/github.com/imdario/mergo/map.go:75]: @liggitt  So, how about this PR? shall we upgrade the mergo version preserving existing behavior now? Or, wait for mergo project fixing it then upgrade mergo?

[the remark this answers is not in the document]
Comment (liggitt) [reply vendor/github.com/imdario/mergo/map.go:75]: I think I'd rather see us drop mergo and implement our own merging of these fields given the repeated behavior reversals and long-standing bugs. Our usage seems very limited

**Comment by liggitt** [comment]
/close

I don't see us picking up a version of mergo with the bug noted in https://github.com/kubernetes/kubernetes/pull/107837#discussion_r815940875

I think the best resolution of https://github.com/kubernetes/kubernetes/issues/107499, given the number of times this dependency has reversed behavior on minor bumps, would be to eliminate use of the dependency and write our own merge of our narrow use

**Pull request closed without merging**